woocommerce/woocommerce
# Low stock threshold for variations

### All Submissions:






### Changes proposed in this Pull Request:



Closes #22706.
Closes #27371.

This PR adds support for low stock threshold for variations.

It needs a bit of documentation to be added to our public user-facing docs.

In principle, I tried to keep the changes to the minimum. 
I've added a field to the variation admin UI to allow setting up the low stock amount iff the variation is managing its stock:
![Screenshot 2021-03-11 at 16 25 06](https://user-images.githubusercontent.com/2207451/110810985-96abc480-8286-11eb-99a0-f02ef7a8a852.png)

UPDATE: 

If the value is not set, it falls back to the parent product's setting (if it's set to manage stock) and only then to store-wide setting (you might notice that the number is a slightly lighter shade than other inputs):
![Screenshot 2021-03-11 at 16 27 31](https://user-images.githubusercontent.com/2207451/110811216-c8bd2680-8286-11eb-963d-77738d6f55fa.png)

The setting comes from WC > Product > Inventory settings:
![Screenshot 2021-03-11 at 16 29 32](https://user-images.githubusercontent.com/2207451/110811465-00c46980-8287-11eb-8c08-3137e8ecda68.png)

If the low stock amount is set both on the parent variable product (manage stock is on, low stock amount set to X) and on the variation (also manage stock is on, low stock amount set to Y), the amount from the variation (Y) gets used. If there is no value set on the variation, the amount X from the parent product gets used.

I also enabled setting the **variation low stock amount for all variations at once** from admin:
![Screenshot 2021-03-11 at 16 36 15](https://user-images.githubusercontent.com/2207451/110812482-f060be80-8287-11eb-896c-2d403c485bae.png)

I was a bit surprised to see that the **Low stock report** only checks the stock level vs the store-wide setting (WC > Product > Inventory), so not much to do here really, unless we change the report (which we don't want to do anyway, as it's part of the reports that are being deprecated).

I checked the **Low stock notification email** gets sent correctly when enabled and shows info about the variation:
![Screenshot 2021-03-11 at 16 38 45](https://user-images.githubusercontent.com/2207451/110812889-4a618400-8288-11eb-9940-4020f7921ee3.png)

Additionally, I checked the **CSV Import and Export**, both seem to work fine with this PR.

Finally, I've added the field to the v3 REST API both for GET and POST (to resolve #27371).

Side note: REST API docs need to be regenerated to make sure it includes the new fields.

### How to test the changes in this Pull Request:

1. Create a variation product (P1) with at least 3 variations, set 2 variations to manage stock, one not to. Parent product set to NOT manage stock.
2. Create a variation product (P2) with at least 3 variations, set 2 variations to manage stock, one not to. Parent product set to manage stock.
2. Apply branch.
2. Set the P1.V1 low stock amount to X (different to store-wide setting A), leave the P1.V2 low stock amount to no value (so that it takes the default value from store-wide setting). Set the same for P2 variations, while also setting the product P2's low stock amount to value Z.
3. Set the stock above the limits for these products.
4. Submit orders for the different variations and see if you observe expected behavior.
5. Check that the emails are being sent at the correct stock level for each variation.
6. Export both products, import them back as new products and confirm that the low stock amounts are set correctly.
7. Try to set the low stock amount for all variations via the Bulk Actions dropdown.
8. Check the REST API response for v2 GET requests `http://one.wordpress.test/wp-json/wc/v2/products/P1/` and `http://one.wordpress.test/wp-json/wc/v2/products/P1/variations/V1` to check there are no changes. 
9. Check the REST API response for v3 GET requests `http://one.wordpress.test/wp-json/wc/v3/products/P1/` and `http://one.wordpress.test/wp-json/wc/v3/products/P1/variations/V1` and check if the correct responses get returned, including new data.
9. Try to update the product and variation low stock amounts using a POST request with request body such as `{ "low_stock_amount": 7 }`

### Other information:




### Changelog entry

> Enhancement - Add support for the low stock threshold for variations.
